# HG changeset patch
# User Lin Ming <mlin@ss.pku.edu.cn>
# Date 1333384359 -3600
# Node ID 6f81f4d79fde3fe95a67545267567cd825169ce8
# Parent  86f9fd67152ba4c5bb3870f9b95ae304809eab86
libxl: support for "rtc_timeoffset" and "localtime"

Implement "rtc_timeoffset" and "localtime" options compatible as xm.

rtc_timeoffset is the offset between host time and guest time.
localtime means to specify whether the emulted RTC appears as UTC or is
offset by the host.

Acked-by: Ian Campbell <ian.campbell@citrix.com>
Signed-off-by: Lin Ming <mlin@ss.pku.edu.cn>
Committed-by: Ian Jackson <ian.jackson.citrix.com>
Backported-to-4-1-by: Marek Marczykowski <marmarek@invisiblethingslab.com>
Index: xen-hvm/tools/libxl/libxl.idl
===================================================================
--- xen-hvm.orig/tools/libxl/libxl.idl
+++ xen-hvm/tools/libxl/libxl.idl
@@ -94,6 +94,8 @@ libxl_domain_build_info = Struct("domain
     ("target_memkb",    uint32),
     ("video_memkb",     uint32),
     ("shadow_memkb",    uint32),
+    ("rtc_timeoffset",  uint32),
+    ("localtime",       bool),
     ("disable_migrate", bool),
     ("kernel",          libxl_file_reference),
     ("cpuid",           libxl_cpuid_policy_list),
Index: xen-hvm/tools/libxl/libxl_create.c
===================================================================
--- xen-hvm.orig/tools/libxl/libxl_create.c
+++ xen-hvm/tools/libxl/libxl_create.c
@@ -83,6 +83,8 @@ void libxl_init_build_info(libxl_domain_
     b_info->cur_vcpus = 1;
     b_info->max_memkb = 32 * 1024;
     b_info->target_memkb = b_info->max_memkb;
+	b_info->localtime = 0;
+	b_info->rtc_timeoffset = 0;
     b_info->disable_migrate = 0;
     b_info->cpuid = NULL;
     b_info->shadow_memkb = 0;
Index: xen-hvm/tools/libxl/libxl_dom.c
===================================================================
--- xen-hvm.orig/tools/libxl/libxl_dom.c
+++ xen-hvm/tools/libxl/libxl_dom.c
@@ -76,6 +76,19 @@ int libxl__build_pre(libxl_ctx *ctx, uin
     if ( info->disable_migrate )
         xc_domain_disable_migrate(ctx->xch, domid);
 
+	if (info->localtime) {
+        time_t t;
+        struct tm *tm;
+
+        t = time(NULL);
+        tm = localtime(&t);
+
+        info->rtc_timeoffset += tm->tm_gmtoff;
+	}
+
+    if (info->rtc_timeoffset)
+        xc_domain_set_time_offset(ctx->xch, domid, info->rtc_timeoffset);
+
     if (info->hvm) {
         unsigned long shadow;
         shadow = (info->shadow_memkb + 1023) / 1024;
Index: xen-hvm/tools/libxl/xl_cmdimpl.c
===================================================================
--- xen-hvm.orig/tools/libxl/xl_cmdimpl.c
+++ xen-hvm/tools/libxl/xl_cmdimpl.c
@@ -749,6 +749,12 @@ static void parse_config_data(const char
     if (!xlu_cfg_get_long(config, "tsc_mode", &l))
         b_info->tsc_mode = l;
 
+    if (!xlu_cfg_get_long(config, "rtc_timeoffset", &l))
+        b_info->rtc_timeoffset = l;
+
+    if (!xlu_cfg_get_long (config, "localtime", &l))
+		b_info->localtime = l;
+
     if (!xlu_cfg_get_long (config, "videoram", &l))
         b_info->video_memkb = l * 1024;
 
